routing/payment_session.go

```diff
@@ -332,14 +328,38 @@ func (p *paymentSession) RequestRoute(maxAmt, feeLimit lnwire.MilliSatoshi,
 			p.selfNode, p.selfNode, p.payment.Target,
 			maxAmt, p.payment.TimePref, finalHtlcExpiry,
 		)
+		if err != nil {
+			// Wrap the error to distinguish path finding errors
+			// from other errors in this closure.
+			return fmt.Errorf("%w: %w", errPathFinding, err)
+		}
 
-		// Close routing graph session.
-		if err := closeGraph(); err != nil {
-			log.Errorf("could not close graph session: %v", err)
+		return nil
+	}
+
+	for {
+		// Get a routing graph session.
+		graph, closeGraph, err := p.graphSessFactory.NewGraphSession()
+		if err != nil {
+			return nil, err
+		}
+
+		err = findPath(graph)
+		// First, close routing graph session.
+		// NOTE: this will be removed in an upcoming commit.
+		if graphErr := closeGraph(); graphErr != nil {
+			log.Errorf("could not close graph session: %v",
+				graphErr)
+		}
+		// If there is an error, and it is not a path finding error, we
+		// return it immediately.
+		if err != nil && !errors.Is(err, errPathFinding) {
+			return nil, err
 		}
 
+		// Otherwise, we'll switch on the path finding error.
```

Comment:
At this point we're sure that it's an `errPathFinding` error so we might as well `Unwrap` it so that we can avoid unnecessary `errors.Is` calls in the switch.

Reply:
dont we anyways want `errors.Is` below so that we dont have to worry about our pathfinding function using error wrapping if it changes from how it stands today?

Reply:
I was wondering of the impact so made a little benchmark: https://go.dev/play/p/VLzVe2GUDrL

```
cpu: Apple M2 Pro
BenchmarkErrorsIs-12          	13187875	        91.21 ns/op	      56 B/op	       2 allocs/op
BenchmarkDirectEquality-12    	622278074	         1.930 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/bhandras/tmp	3.879s
```

While error wrapping and unwrapping is costly I think overall would not impact the pathfinding loop. If we experience measurable change that affects users we can always make it faster :)

Reply:
i guess this is one area where we really do want to optimise for speed though right? 

so can defs change it back - would just need to implement an "Unwrap"-able type but can defs do that.

Reply:
updated!

Reply:
LGTM!

Reply:
I don't think the benchmark is correct - it should compare `error.Is` vs `error.As` but not a direct `==`? Also we should separate benchmarking the creation of wrapped errors vs the unwrapping of errors.

Reply:
> but not a direct ==

but direct `==` is what we are talking about? see the error comparisions below